Omitempty attribute causes compiler error on field type struct

### Describe the bug

Omitempty attribute on a field that is a struct causes compiler error.

### Expected Behavior

Expected omitempty attribute on a struct field to omit empty struct when encoding to json.

### Current Behavior

```
/tmp/v_501/omitempty.7550253850255435104.tmp.c:2242:14: error: invalid operands to binary expression ('main__Bug' (aka 'struct main__Bug') and 'main__Bug')
        if (val.bug != (main__Bug){.name = (string){.str=(byteptr)"", .is_lit=1},})
            ~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps


```
struct OmitEmptyBug {
	bug Bug [omitempty]
}

struct Bug {
	name string
}

fn main() {
	test := OmitEmptyBug {
		bug: Bug{
			name: 'mybug'
		}
	}
	encoded := json.encode(test)
}
```

### Possible Solution

If omitempty isn't yet supported on fields that are structs, a better error message can be printed.

### Additional Information/Context

_No response_

### V version

0.3.3

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 f1e9a8f.adcd16b
OS: macos, macOS, 13.2.1, 22D68
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz

--- a/vlib/json/json_omitempty_types_test.v
+++ b/vlib/json/json_omitempty_types_test.v
@@ -1,0 +1,104 @@
+import json
+
+struct OmitEmptyStruct {
+	bug Struct [omitempty]
+}
+
+type MyAlias = string
+
+struct OmitEmptyAlias {
+	bug MyAlias [omitempty]
+}
+
+struct Struct {
+	name string
+}
+
+struct OmitEmptyMap {
+	bug map[string]string [omitempty]
+}
+
+struct OmitEmptyArray {
+	bug []string [omitempty]
+}
+
+type MySum = int | string
+
+struct OmitEmptySumType {
+	bug MySum [omitempty]
+}
+
+fn main() {
+	test_struct()
+	test_alias()
+	test_map()
+	test_sumtype()
+	test_array()
+}
+
+fn test_struct() {
+	test2 := OmitEmptyStruct{
+		bug: Struct{}
+	}
+	encoded2 := json.encode(test2)
+	dump(encoded2)
+	test := OmitEmptyStruct{
+		bug: Struct{
+			name: 'mybug'
+		}
+	}
+	encoded := json.encode(test)
+	dump(encoded)
+}
+
+fn test_alias() {
+	test3 := OmitEmptyAlias{
+		bug: ''
+	}
+	encoded3 := json.encode(test3)
+	dump(encoded3)
+	test4 := OmitEmptyAlias{
+		bug: 'foo'
+	}
+	encoded4 := json.encode(test4)
+	dump(encoded4)
+}
+
+fn test_map() {
+	test3 := OmitEmptyMap{
+		bug: {}
+	}
+	encoded3 := json.encode(test3)
+	dump(encoded3)
+	test4 := OmitEmptyMap{
+		bug: {
+			'foo': 'bar'
+		}
+	}
+	encoded4 := json.encode(test4)
+	dump(encoded4)
+}
+
+fn test_sumtype() {
+	test3 := OmitEmptySumType{}
+	encoded3 := json.encode(test3)
+	dump(encoded3)
+	test4 := OmitEmptySumType{
+		bug: 1
+	}
+	encoded4 := json.encode(test4)
+	dump(encoded4)
+}
+
+fn test_array() {
+	test3 := OmitEmptyArray{
+		bug: []
+	}
+	encoded3 := json.encode(test3)
+	dump(encoded3)
+	test4 := OmitEmptyArray{
+		bug: ['1']
+	}
+	encoded4 := json.encode(test4)
+	dump(encoded4)
+}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -681,17 +681,19 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 			} else if field.typ == ast.string_type {
 				enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)}.len != 0)')
 			} else {
-				ptr_typ := g.equality_fn(field.typ)
-				if field_sym.kind == .alias {
-					enc.writeln('${indent}if (!${ptr_typ}_alias_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
-				} else if field_sym.kind == .sum_type {
-					enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)}._typ != 0)')
-				} else if field_sym.kind == .map {
-					enc.writeln('${indent}if (!${ptr_typ}_map_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
-				} else if field_sym.kind == .array {
-					enc.writeln('${indent}if (!${ptr_typ}_arr_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
-				} else if field_sym.kind == .struct_ {
-					enc.writeln('${indent}if (!${ptr_typ}_struct_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+				if field_sym.kind in [.alias, .sum_type, .map, .array, .struct_] {
+					ptr_typ := g.equality_fn(field.typ)
+					if field_sym.kind == .alias {
+						enc.writeln('${indent}if (!${ptr_typ}_alias_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+					} else if field_sym.kind == .sum_type {
+						enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)}._typ != 0)')
+					} else if field_sym.kind == .map {
+						enc.writeln('${indent}if (!${ptr_typ}_map_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+					} else if field_sym.kind == .array {
+						enc.writeln('${indent}if (!${ptr_typ}_arr_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+					} else if field_sym.kind == .struct_ {
+						enc.writeln('${indent}if (!${ptr_typ}_struct_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+					}
 				} else {
 					enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)} != ${g.type_default(field.typ)})')
 				}

--- a/vlib/v/gen/c/json.v
+++ b/vlib/v/gen/c/json.v
@@ -681,7 +681,20 @@ fn (mut g Gen) gen_struct_enc_dec(utyp ast.Type, type_info ast.TypeInfo, styp st
 			} else if field.typ == ast.string_type {
 				enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)}.len != 0)')
 			} else {
-				enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)} != ${g.type_default(field.typ)})')
+				ptr_typ := g.equality_fn(field.typ)
+				if field_sym.kind == .alias {
+					enc.writeln('${indent}if (!${ptr_typ}_alias_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+				} else if field_sym.kind == .sum_type {
+					enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)}._typ != 0)')
+				} else if field_sym.kind == .map {
+					enc.writeln('${indent}if (!${ptr_typ}_map_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+				} else if field_sym.kind == .array {
+					enc.writeln('${indent}if (!${ptr_typ}_arr_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+				} else if field_sym.kind == .struct_ {
+					enc.writeln('${indent}if (!${ptr_typ}_struct_eq(${prefix_enc}${op}${c_name(field.name)}, ${g.type_default(field.typ)}))')
+				} else {
+					enc.writeln('${indent}if (${prefix_enc}${op}${c_name(field.name)} != ${g.type_default(field.typ)})')
+				}
 			}
 		}
 		if !is_js_prim(field_type) {

